### PR TITLE
Fix 135

### LINKED ISSUE #135: Scheduling a short setTimeout followed by a long one, causes the short one to hang until the long one executes
Consider the case where we call `setTimeout(0)` and then `setTimeout(500)`
Currently if the `setTimeout(500)` is called in the same run loop as the first one, it will cause the `setTimeout(0)` to also wait 500 ms.

This bug was introduced by https://github.com/ebryn/backburner.js/pull/94

In particular, https://github.com/ebryn/backburner.js/pull/94/files#diff-820bd966639ebce8fcbffd97fb5dac41R441seems wrong, because if we have an expired timeout that hasn't fired, `wait` time should be `0`. 
I am a bit concerned with the computer sleep issue, because now and also with this fix, you could get  a timeout from couple days ago which is not how the native browser `setTimeout` behaves. Maybe we should have a max expired time after which we ignore the timeouted function?

This seems like a probable cause of https://github.com/emberjs/ember.js/issues/10479

### PR DESCRIPTION
**Pull request by krisselden**
some of the tests intermittently fail.

Closes #146
Closes #137
Closes #143
Fixes #135

### CODE REVIEW

File: lib/backburner.js [hunk 1]
```diff
@@ -30,6 +30,16 @@ export default function Backburner(queueNames, options) {
     end: [],
     begin: []
   };
+
+  var backburner = this;
+  this.executeTimers = function () {
```
Comment (stefanpenner) [inline lib/backburner.js]: Why not just put this on the prototype

File: lib/backburner.js [hunk 2]
```diff
@@ -30,6 +30,16 @@ export default function Backburner(queueNames, options) {
...
+    backburner._laterTimer = null;
+    backburner._laterTimerExpiresAt = null;
+    backburner.run(backburner, executeTimers);
+    if (backburner._timers.length) {
```
Comment (stefanpenner) [inline lib/backburner.js]: This feels somewhat risky as an error above may prevent this from running

File: lib/backburner.js (hunk 1 above)
Comment (krisselden) [inline lib/backburner.js]: Because the point is to bind to this so we don't remake the bound function every time.

File: lib/backburner.js (hunk 2 above)
Comment (krisselden) [inline lib/backburner.js]: I don't think I changed that aspect of the code but noted

Comment (stefanpenner) [inline lib/backburner.js]: its likely worth adding a try/finally here.

Comment (krisselden) [inline lib/backburner.js]: I can move it to after they are scheduled but before the loop is flushed.

File: lib/backburner.js (hunk 1 above)
Comment (stefanpenner) [inline lib/backburner.js]: are you ok with moving the content of this function to on the prototype?

```js
prototype.executeTimers = function () {
  .. content of your function.
}
```

We can then in the constructor do what you just did.

```js
var backburner = this;
this.boundExecuteTimers = function() {
  return backburner.executeTimers.bind(backburner);
}
```

Why?

* the function logic is where one would expect it to be.

**Comment by krisselden** [comment]
@rwjblue @stefanpenner when tests are green this is merge-able pending review.

**Comment by stefanpenner** [comment]
tests appear unhappy

**Comment by rwjblue** [comment]
LGTM 

@stefanpenner - r?

**Comment by stefanpenner** [comment]
LGTM

**Comment by stefanpenner** [comment]
following up with another quick cleanup PR

**Comment by stefanpenner** [comment]
follow up pr https://github.com/ebryn/backburner.js/pull/152

**Pull request merged**